transport/mixer/src/channel.rs

```diff
@@ -0,0 +1,469 @@
+use futures::{Future, SinkExt, Stream, StreamExt};
+use futures_timer::Delay;
+use std::{
+    cmp::Reverse,
+    collections::BinaryHeap,
+    future::poll_fn,
+    sync::{Arc, Mutex},
+    task::Poll,
+    time::Duration,
+};
+use tracing::trace;
+
+use crate::{config::MixerConfig, data::DelayedData};
+
+#[cfg(all(feature = "prometheus", not(test)))]
+use hopr_metrics::metrics::SimpleGauge;
+
+#[cfg(all(feature = "prometheus", not(test)))]
+lazy_static::lazy_static! {
+    static ref METRIC_QUEUE_SIZE: SimpleGauge =
+        SimpleGauge::new("hopr_mixer_queue_size", "Current mixer queue size").unwrap();
+    static ref METRIC_MIXER_AVERAGE_DELAY: SimpleGauge = SimpleGauge::new(
+        "hopr_mixer_average_packet_delay",
+        "Average mixer packet delay averaged over a packet window"
+    )
+    .unwrap();
+}
+
+/// Mixing and delaying channel using random delay function.
+///
+/// Mixing is performed by assigning random delays to the ingress timestamp of data,
+/// then storing the values inside a binary heap with reversed ordering (max heap).
+/// This effectively creates a min heap behavior, which is required to ensure that
+/// data is released in order of their delay expiration.
+///
+/// When data arrives:
+/// 1. A random delay is assigned
+/// 2. Data is stored in the heap with its release timestamp
+/// 3. The heap maintains ordering so items with earliest release time are at the top
+///
+/// The channel uses a single timer thread that is instantiated on the first
+/// timer reset and shared across all operations. This channel is **unbounded** by nature
+/// using the `capacity` in the configuration to solely pre-allocate the buffer.
+struct Channel<T> {
+    /// Buffer holding the data with a timestamp ordering to ensure the min heap behavior.
+    buffer: BinaryHeap<Reverse<DelayedData<T>>>,
+    timer: futures_timer::Delay,
+    waker: Option<std::task::Waker>,
+    sender_count: usize,
+    receiver_active: bool,
+    cfg: MixerConfig,
+}
+
+/// Error returned by the [`Sender`].
+#[derive(Debug, thiserror::Error)]
+pub enum SenderError {
+    /// The channel is closed due to receiver being dropped.
+    #[error("Channel is closed")]
+    Closed,
+
+    /// The mutex lock over the channel failed.
+    #[error("Channel lock failed")]
+    Lock,
+}
+
+/// Sender object interacting with the mixing channel.
+pub struct Sender<T> {
+    channel: Arc<Mutex<Channel<T>>>,
+}
+
+impl<T> Clone for Sender<T> {
+    fn clone(&self) -> Self {
+        let mut channel = self.channel.lock().unwrap();
+        channel.sender_count += 1;
+
+        Sender {
+            channel: self.channel.clone(),
+        }
+    }
+}
+
+impl<T> Drop for Sender<T> {
+    fn drop(&mut self) {
+        let mut channel = self.channel.lock().unwrap();
```

Comment:
`drop` should not panic, see: https://doc.rust-lang.org/std/ops/trait.Drop.html#panics

In this case, it would be safe to track on Sender's own `is_closed` member and checking the `Arc::strong_count(self.channel)` instead. If that is equal to `2`, meaning it there's only the last Sender and the Receiver itself, the poison on the Mutex should be cleared (via `clear_poison`) and then the mutex can be safely unwrapped and the Channel closed.

Reply:
What if the receiver is dropped with 2 senders being active?

Reply:
Can you in addition use `Arc<AtomicBool>` to track that the Receiver has  been dropped ?

Reply:
Sure, I'd track everything with atomics for sender and receiver tracking.